googleapis/rules/rule_lib.py

```diff
@@ -0,0 +1,457 @@
+
+    Returns:
+      {'name': 'operations/rulejob_jo_0440fca2-ff4b-4067-b16a-e8bc2368fc15'}
+    """
+    if not rule_id or not start_time or not end_time:
+      raise ValueError(
+          "Missing one or more of rule id, start time, or end time")
+
+    url = "{}/rules/{}:run".format(self.backstory_api_url, rule_id)
+    _LOGGER_.info("run rule: %s ", url)
+
+    # Construct the post body for the create rule request.
+    body = {
+        "event_start_time": start_time,
+        "event_end_time": end_time
+    }
+    # Make a request
+    response = self.client.request(
+        url,
+        method="POST",
+        headers={"Content-type": "application/x-www-form-urlencoded"},
+        body=urllib.parse.urlencode(body))
+    return self._parse_response(response)
+
+  def list_results(self, operation_id, page_size=0, page_token=''):
+    """Call the ListResults RPC.
+
+    Args:
+      operation_id: an operation id
+
+    Returns:
+      {'metadata': {
+          'eventEndTime': '2019-12-04T17:02:47Z',
+          'ruleId': 'ru_13f58277-8d41-45b2-ab6b-1addc4e3ca0e',
+          'eventStartTime': '2019-11-25T00:00:00Z'
+          },
+       'results': [
+             {'match': {
+               ... sinlge event in udm format ...
+                       }
+              },
+           ...
+           ]
+      }
+    """
+    if not operation_id:
+      raise ValueError("Missing operation id")
+    url = "{}/rules_results?name={}&page_size={}&page_token={}".format(
+        self.backstory_api_url, operation_id, page_size, page_token)
+    _LOGGER_.info("list results: %s ", url)
+
+    response = self.client.request(url, "GET")
+    return self._parse_response(response)
+
+  def get_operation(self, operation_id):
+    """Call the GetOperation RPC.
+
+    Args:
+      operation_id: an operation id
+
+    Returns:
+      {'done': True,
+         'error': {'code': 10, 'message': 'RESULT_LIMIT_REACHED'},
+         'metadata': {'@type': 'type.googleapis.com/chronicle.backstory.v1.RunRuleMetadata',
+                      'eventEndTime': '2019-11-25T00:00:00Z',
+                      'ruleId': 'ru_d1b69671-10d8-451a-9398-b6a254aa22e0',
+                      'eventStartTime': '2019-11-11T17:31:59Z'},
+         'name': 'operations/rulejob_jo_1b6c6267-53ed-4667-819b-a024ea8bfbe2'
+      }
+    """
+    if not operation_id:
+      raise ValueError("Missing operation id")
+    url = "{}/{}".format(self.backstory_api_url, operation_id)
+    _LOGGER_.info("get operations: %s ", url)
+
+    # Make a request
+    response = self.client.request(url, "GET")
+    return self._parse_response(response)
+
+  def list_operations(self):
+    """Call the ListOperations RPC.
+
+    Sadly, done is only populated if the job is done. Not done needs to be
+    inferred. Similar for success which needs to be inferred by a
+    lack of error. Not clear why eventStartTime and eventEndTime is in some of
+    them and not others.
+
+    Returns:
+      [ {'done': True,
```

Comment:
What are your thoughts on making this one "False" to demonstrate an operation that is still running?

Reply:
I think not, since a still-running op will AFAIK not set 'done' at all.  Maybe good to add a second example, though, to cover this.